# Introduction

The National Longitudinal Survey of Youth 1997-2011 dataset is one of the most important databases available to social scientists working with US data. 

It allows scientists to look at the determinants of earnings as well as educational attainment and has incredible relevance for government policy. It can also shed light on politically sensitive issues like how different educational attainment and salaries are for people of different ethnicity, sex, and other factors. When we have a better understanding how these variables affect education and earnings we can also formulate more suitable government policies. 

<center><img src=https://i.imgur.com/cxBpQ3I.png height=400></center>


### Upgrade Plotly

In [78]:
%pip install --upgrade plotly

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


###  Import Statements


In [79]:
import pandas as pd
import numpy as np

import seaborn as sns
import plotly.express as px
import matplotlib.pyplot as plt

from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

## Notebook Presentation

In [80]:
pd.options.display.float_format = '{:,.2f}'.format

# Load the Data



In [81]:
df_data = pd.read_csv('NLSY97_subset.csv')

### Understand the Dataset

Have a look at the file entitled `NLSY97_Variable_Names_and_Descriptions.csv`. 

---------------------------

    :Key Variables:  
      1. S           Years of schooling (highest grade completed as of 2011)
      2. EXP         Total out-of-school work experience (years) as of the 2011 interview.
      3. EARNINGS    Current hourly earnings in $ reported at the 2011 interview

In [82]:

# df_complete_data = pd.read_csv('NLSY97_Variable_Names_and_Descriptions.csv')
# df_complete_data.info()

In [83]:
# df_complete_data.head()

# Preliminary Data Exploration 🔎

**Challenge**

* What is the shape of `df_data`? 
* How many rows and columns does it have?
* What are the column names?
* Are there any NaN values or duplicates?

In [84]:
df_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 96 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   ID        2000 non-null   int64  
 1   EARNINGS  2000 non-null   float64
 2   S         2000 non-null   int64  
 3   EXP       2000 non-null   float64
 4   FEMALE    2000 non-null   int64  
 5   MALE      2000 non-null   int64  
 6   BYEAR     2000 non-null   int64  
 7   AGE       2000 non-null   int64  
 8   AGEMBTH   1956 non-null   float64
 9   HHINC97   1630 non-null   float64
 10  POVRAT97  1627 non-null   float64
 11  HHBMBF    2000 non-null   int64  
 12  HHBMOF    2000 non-null   int64  
 13  HHOMBF    2000 non-null   int64  
 14  HHBMONLY  2000 non-null   int64  
 15  HHBFONLY  2000 non-null   int64  
 16  HHOTHER   2000 non-null   int64  
 17  MSA97NO   2000 non-null   int64  
 18  MSA97NCC  2000 non-null   int64  
 19  MSA97CC   2000 non-null   int64  
 20  MSA97NK   2000 non-null   int6

In [85]:
df_data.head()

ID  EARNINGS   S  EXP  FEMALE  MALE  BYEAR  AGE  AGEMBTH   HHINC97  ...  \
0  4275     18.50  12 9.71       0     1   1984   27    24.00 64,000.00  ...   
1  4328     19.23  17 5.71       0     1   1982   29    32.00  6,000.00  ...   
2  8763     39.05  14 9.94       0     1   1981   30    23.00 88,252.00  ...   
3  8879     16.80  18 1.54       0     1   1983   28    30.00       NaN  ...   
4  1994     36.06  15 2.94       0     1   1984   27    23.00 44,188.00  ...   

   URBAN  REGNE  REGNC  REGW  REGS  MSA11NO  MSA11NCC  MSA11CC  MSA11NK  \
0      1      0      0     1     0        0         0        1        0   
1      2      0      0     1     0        0         1        0        0   
2      1      0      0     0     1        0         0        1        0   
3      1      0      1     0     0        0         1        0        0   
4      1      0      0     0     1        0         0        1        0   

   MSA11NIC  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 96 columns]

## Data Cleaning - Check for Missing Values and Duplicates

Find and remove any duplicate rows.

In [86]:
print(f"Number of duplicate rows: {df_data.duplicated().values.sum()}")

Number of duplicate rows: 513


In [87]:
df_data.drop_duplicates(inplace=True)

In [88]:
df_data.dropna(inplace=True)

## Descriptive Statistics

In [89]:
df_data.describe()

ID  EARNINGS      S    EXP  FEMALE   MALE    BYEAR    AGE  \
count   492.00    492.00 492.00 492.00  492.00 492.00   492.00 492.00   
mean  3,530.57     19.13  14.89   5.92    0.49   0.51 1,982.98  28.02   
std   1,948.08     11.54   2.69   2.51    0.50   0.50     0.82   0.82   
min      28.00      2.13   8.00   0.00    0.00   0.00 1,982.00  27.00   
25%   1,833.25     12.00  12.00   4.24    0.00   0.00 1,982.00  27.00   
50%   3,470.50     16.00  16.00   5.75    0.00   1.00 1,983.00  28.00   
75%   5,186.75     24.04  17.00   7.75    1.00   1.00 1,984.00  29.00   
max   8,978.00    123.08  20.00  12.33    1.00   1.00 1,984.00  29.00   

       AGEMBTH    HHINC97  ...  URBAN  REGNE  REGNC   REGW   REGS  MSA11NO  \
count   492.00     492.00  ... 492.00 492.00 492.00 492.00 492.00   492.00   
mean     26.74  66,732.78  ...   0.75   0.13   0.31   0.35   0.21     0.05   
std       4.71  44,951.87  ...   0.44   0.33   0.46   0.48   0.41     0.22   
min      17.00       0.00  ...   0.00   0.00   0.00   0.00   0.00     0.00   
25%      24.00  40,725.00  ...   0.00   0.00   0.00   0.00   0.00     0.00   
50%      26.00  58,027.50  ...   1.00   0.00   0.00   0.00   0.00     0.00   
75%      30.00  77,432.50  ...   1.00   0.00   1.00   1.00   0.00     0.00   
max      41.00 246,474.00  ...   2.00   1.00   1.00   1.00   1.00     1.00   

       MSA11NCC  MSA11CC  MSA11NK  MSA11NIC  
count    492.00   492.00   492.00    492.00  
mean       0.54     0.41     0.00      0.00  
std        0.50     0.49     0.05      0.00  
min        0.00     0.00     0.00      0.00  
25%        0.00     0.00     0.00      0.00  
50%        1.00     0.00     0.00      0.00  
75%        1.00     1.00     0.00      0.00  
max        1.00     1.00     1.00      0.00  

[8 rows x 96 columns]

## Visualise the Features

In [90]:
fig = px.scatter(df_data, y='EARNINGS', x='HHINC97', title='Earnings now vs Household income prior 97', trendline='lowess')
fig.show()

Not much of a trend between earnings and Household income prior 1997.

In [91]:
columns =['CATGOV','CATPRI', 'CATSE','CATNPO']
emp_cat = []
for col in columns:
    emp_cat.append(np.count_nonzero(df_data[col]))

px.bar(x=columns, y=emp_cat, title= 'Category of Earnings')
    


Most people are employed in Private sector

# Split Training & Test Dataset

We *can't* use all the entries in our dataset to train our model. Keep 20% of the data for later as a testing dataset (out-of-sample data).  

In [92]:
target = df_data['EARNINGS']
features = df_data.drop(['EARNINGS','ASVABC', 'ASVABC4', 'VERBAL'], axis=1)

X_train, X_test, y_train, y_test = train_test_split(features, target, 
                                                    test_size=0.2, 
                                                    random_state=3)

Earnings along with few other variables that had high correlation were removed from features.

TRAINING & TEST DATA FOR UNIVARIATE DISTRIBUTION OF EARNINGS AGAINST SCHOOLING

In [106]:
targetS = df_data['EARNINGS']
featuresS = df_data['S']

X_trainS, X_testS, y_trainS, y_testS = train_test_split(featuresS, targetS, 
                                                    test_size=0.2, 
                                                    random_state=3)

# Simple Linear Regression

Only use the years of schooling to predict earnings. Use sklearn to run the regression on the training dataset. How high is the r-squared for the regression on the training data? 

Regression with all Variables

In [111]:
regr = LinearRegression()
regr.fit(X_train, y_train)
rsquared = regr.score(X_train, y_train)

print(f"{rsquared:.2}")


0.4


Regression with ony Schooling as the independent variable

In [108]:
X_trainS = X_trainS.values.reshape(-1,1)
X_testS = X_testS.values.reshape(-1,1)

In [109]:
regrS = LinearRegression()
regrS.fit(X_trainS, y_trainS)
rsquaredS = regrS.score(X_trainS, y_trainS)

print(f"{rsquaredS:.2}")


0.081


Based on the R2 value, Earnings are poorly predicted by Schooling. 

### Evaluate the Coefficients of the Model

Here we do a sense check on our regression coefficients. The first thing to look for is if the coefficients have the expected sign (positive or negative). 

Interpret the regression. How many extra dollars can one expect to earn for an additional year of schooling?

Based on the univariate distribution, the regression coefficient is below.
Each additional year of schooling will increase by $1.2 per hour

In [112]:
regrS.coef_

array([1.19823989])

In [113]:
coeff_df = (pd.DataFrame({'coef':regr.coef_, 'category':features.columns})
.sort_values(by = 'coef', ascending = False)
.set_index(['category']))

In [99]:
coeff_df.tail()

coef
category      
EDUCGED  -3.87
REGNE    -3.91
EDUCHSD  -4.30
EDUCDO   -5.28
HHOTHER  -6.86

In [100]:
# coeff_df.iat[i,coeff_df.columns.get_loc('S')]
i = coeff_df.index.get_loc('S')
print(f'Additional earnings for every additional year of schooling {coeff_df.values[i][0]:.2} per hour')

Additional earnings for every additional year of schooling 0.36 per hour


### Analyse the Estimated Values & Regression Residuals

How good our regression is also depends on the residuals - the difference between the model's predictions ( 𝑦̂ 𝑖 ) and the true values ( 𝑦𝑖 ) inside y_train. Do you see any patterns in the distribution of the residuals?

In [114]:
y_capS = regrS.predict(X_trainS)
residualS = y_trainS - y_capS

In [115]:
fig = px.scatter(x=y_trainS, y=y_capS, title='Actual vs Predicted Earnings(Based on schooling alone)', 
                 labels={'x':'Actual', 'y': 'Predicted'}, trendline='ols')
fig.show()

In [116]:
fig = px.scatter(x=y_trainS, y=residualS, title='Actual vs Residuals (Univariate - Schooling)', 
                 labels={'x':'Actual Values', 'y': 'Residuals'},)
fig.show()

Linear distribution between actual and residuals

For Multivariate distribution with all variables

In [72]:
y_cap = regr.predict(X_train)
residual = y_train - y_cap 

In [73]:
fig = px.scatter(x=y_train, y=y_cap, title='Actual vs Predicted values', 
                 labels={'x':'Actual Values', 'y': 'Predicted Values'}, 
                 trendline='ols')
fig.show()

The model seems to underpredict higher values and overpredict smaller values. 

In [74]:
fig = px.scatter(x=y_train, y=residual, title='Actual vs Residuals', 
                 labels={'x':'Actual Values', 'y': 'Residuals'},)
fig.show()

Residuals are increasing with increase in actual values. 

# Multivariable Regression

Now use both years of schooling and the years work experience to predict earnings. How high is the r-squared for the regression on the training data? 

In [119]:
target3 = df_data['EARNINGS']
features3 = df_data[['S', 'EXP']]

X_train3, X_test3, y_train3, y_test3 = train_test_split(features3, target3, 
                                                    test_size=0.2,random_state=3)

In [121]:
regr3 = LinearRegression()
regr3.fit(X_train3, y_train3)
rsquared3 = regr3.score(X_train3, y_train3)

print(f"The R2 of Earnings vs Experience & Schooling: {rsquared3:.2}")

The R2 of Earnings vs Experience & Schooling: 0.11


### Evaluate the Coefficients of the Model

In [122]:
regr3.coef_

array([1.78560288, 0.96031509])

Based on this regression, each year of schooling increases earnings by $1.78 per hour and each additional year of work increases earnings by $0.98 per hour

### Analyse the Estimated Values & Regression Residuals

In [123]:
y_cap3 = regr3.predict(X_train3)
residual3 = y_train3 - y_cap3

In [124]:
fig = px.scatter(x=y_train3, y=y_cap3, title='Actual vs Predicted values (Experience & Schooling years as variable)', 
                 labels={'x':'Actual Values', 'y': 'Predicted Values'}, 
                 trendline='ols')
fig.show()

In [125]:
fig = px.scatter(x=y_train3, y=residual3, 
                 title='Actual vs Residuals (Experience & Schooling years as variable)', 
                 labels={'x':'Actual Values', 'y': 'Residuals'},)
fig.show()

# Use Your Model to Make a Prediction

How much can someone with a bachelors degree (12 + 4) years of schooling and 5 years work experience expect to earn in 2011?

In [130]:
regr3.predict([[16,5]])

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning:

X does not have valid feature names, but LinearRegression was fitted with feature names



array([20.07370081])

According to the linear regression, the expected earnings for someone with 16 years of schooling and 5 years of work experience is $20.07

# Experiment and Investigate Further

Which other features could you consider adding to further improve the regression to better predict earnings? 